### DISCUSSION TITLE
Battery cycles

### DISCUSSION THREAD

**Discussion by danbi** [fl4p/batmon-ha#56] [Ideas]
It would be useful to read the battery (pack) cycle count from the BMS. At least with Daly this is available, and the add-on reads it

INFO     [daly.py:36] got daly states: {'num_cells': 16, 'num_temps': 1, 'charging': False, 'discharging': False, 'states': {'DI1': False, 'DI2': True}, 'num_cycles': 45}

This of course depends on the BMS being properly set up and the value might not be very accurate, but even if not daily/monthly/yearly increase is still useful to monitor.

**Comment by danbi** [comment]
I see the release notes indicate this is now supported, but nothing appears in the dashboard. Instead, there is this in the log:

INFO     [sampling.py:61] connecting bms DalyBt(40:16:06:01:21:E1)
INFO     [main.py:106] mqtt watchdog loop started with timeout 120.0s
WARNING  [daly.py:87] tried rx/tx/sx uuids 17/15/48: Characteristic with UUID 17 could not be found!
INFO     [daly.py:84] found rx uuid to be working: 0000fff1-0000-1000-8000-00805f9b34fb (tx 0000fff2-0000-1000-8000-00805f9b34fb, sx 02f00000-0000-0000-0000-00000000ff01)
INFO     [sampling.py:67] connected bms DalyBt(40:16:06:01:21:E1)!
INFO     [sampling.py:85] battery1 subscribing for {'charge': True, 'discharge': True} switch change
INFO     [mqtt_util.py:304] subscribe homeassistant/switch/battery1/charge/set
INFO     [mqtt_util.py:304] subscribe homeassistant/switch/battery1/discharge/set
INFO     [sampling.py:94] battery1 result@2023-03-03T11:10:09.413646 BmsSample(U=54.9V,I=5.10A,P=280W,q=317.1Ah/320.0,mos=nan°C)
INFO     [daly.py:37] got daly states: {'num_cells': 16, 'num_temps': 1, 'charging': False, 'discharging': False, 'states': {'DI1': False, 'DI2': True}, 'num_cycles': 51}
ERROR    [message_bus.py:683] A message handler raised an exception: invalid state.
Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/dbus_next/message_bus.py", line 668, in _process_message
    result = handler(msg)
  File "/usr/lib/python3.10/site-packages/bleak/backends/bluezdbus/client.py", line 1168, in _parse_msg
    self._notification_callbacks[message.path](
  File "/app/bmslib/daly.py", line 67, in _notification_callback
    self._fetch_futures.set_result(command, response_bytes)
  File "/app/bmslib/__init__.py", line 28, in set_result
    fut.set_result(value)
asyncio.exceptions.InvalidStateError: invalid state
INFO     [sampling.py:103] battery1 volt=3439,3488,3405,3480,3483,3408,3389,3401,3391,3480,3429,3427,3484,3410,3396,3480 temp=[15]
INFO     [sampling.py:128] DalyBt(40:16:06:01:21:E1) times: connect=5.84s fetch=1.21s
/app/bmslib/sampling.py:47: RuntimeWarning: coroutine 'DalyBt.get_states_cached' was never awaited
  return await self.sample()
RuntimeWarning: Enable tracemalloc to get the object allocation traceback

Should I enable more logging?

**Reply by fl4p** [reply]
Appears to be a bug, fixing this with the next update (fix already in master branch here)

**Comment by fl4p** [comment]
Done
